src/graph.rs

```diff
@@ -347,12 +347,79 @@ impl Handle<File> {
 ///
 /// Each node (except for the _root node_ and _jump to scope_ node) lives in a file, and has a
 /// _local ID_ that must be unique within its file.
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[derive(Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd)]
 pub struct NodeID {
-    /// The file that the node comes from.
-    pub file: Handle<File>,
-    /// The unique identity of the node within its file.
-    pub local_id: u32,
+    file: Option<Handle<File>>,
+    local_id: u32,
+}
+
+const ROOT_NODE_ID: u32 = 0;
+const JUMP_TO_NODE_ID: u32 = 1;
+
+impl NodeID {
+    /// Returns the ID of the singleton _root node_.
+    #[inline(always)]
+    pub fn root() -> NodeID {
+        NodeID {
+            file: None,
+            local_id: ROOT_NODE_ID,
+        }
+    }
+
+    /// Returns the ID of the singleton _jump to scope_ node.
+    #[inline(always)]
+    pub fn jump_to() -> NodeID {
+        NodeID {
+            file: None,
+            local_id: JUMP_TO_NODE_ID,
+        }
+    }
+
+    /// Creates a new file-local node ID.
+    #[inline(always)]
+    pub fn new_in_file(file: Handle<File>, local_id: u32) -> NodeID {
+        NodeID {
+            file: Some(file),
+            local_id,
+        }
+    }
+
+    /// Returns whether this ID refers to the singleton _root node_.
+    #[inline(always)]
+    pub fn is_root(self) -> bool {
+        self.file.is_none() && self.local_id == ROOT_NODE_ID
+    }
+
+    /// Returns whether this ID refers to the singleton _jump to scope_ node.
+    #[inline(always)]
+    pub fn is_jump_to(self) -> bool {
+        self.file.is_none() && self.local_id == JUMP_TO_NODE_ID
+    }
+
+    /// Returns the file that this node belongs to.  Returns `None` for the singleton _root_ and
+    /// _jump to scope_ nodes, which belong to all files.
+    #[inline(always)]
+    pub fn file(self) -> Option<Handle<File>> {
+        self.file
+    }
+
+    /// Returns the local ID of this node within its file.  Panics if this node ID refers to the
+    /// singleton _root_ or _jump to scope_ nodes.
+    #[inline(always)]
+    pub fn local_id(self) -> u32 {
+        debug_assert!(self.file.is_some());
```

Comment:
Oh cool, TIL `debug_assert`. For others reading, `debug_assert` is included in non-optimized builds (i.e. `debug`), but is not included in `release` builds without a special flag passed to the compiler.